File: PolymorphicRecursion/Haskell/MySeq.hs
module MySeq (MySeq(..), sizeS) where

data MySeq a = MSNil | MSCons (a, MySeq (a,a))

sizeS :: MySeq a -> Int
sizeS MSNil = 0
sizeS (MSCons (_, ps)) = 1 + 2 * sizeS ps

File: PolymorphicRecursion/OCaml/my_seq.ml
type 'a my_seq = Nil | Cons of 'a * ('a * 'a) my_seq

let rec sizeS : 'a. 'a my_seq -> int = function
  | Nil -> 0
  | Cons (_, my_seq) -> 1 + 2 * sizeS my_seq
